controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackageDiff.java

```diff
@@ -0,0 +1,114 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.yahoo.vespa.hosted.controller.application.pkg.ZipStreamReader.ZipEntryWithContent;
+
+/**
+ * @author freva
+ */
+public class ApplicationPackageDiff {
+
+    public static String diffAgainstEmpty(ApplicationPackage right) {
+        byte[] emptyZip = new byte[]{80, 75, 5, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+        return diff(new ApplicationPackage(emptyZip), right);
+    }
+
+    public static String diff(ApplicationPackage left, ApplicationPackage right) {
+        return diff(left, right, 10 << 20, 1 << 20, 10 << 20);
+    }
+
+    static String diff(ApplicationPackage left, ApplicationPackage right, int maxFileSizeToDiff, int maxDiffSizePerFile, int maxTotalDiffSize) {
+        if (Arrays.equals(left.zippedContent(), right.zippedContent())) return "No diff\n";
+
+        Map<String, ZipEntryWithContent> leftContents = readContents(left, maxFileSizeToDiff);
+        Map<String, ZipEntryWithContent> rightContents = readContents(right, maxFileSizeToDiff);
+
+        StringBuilder sb = new StringBuilder();
+        List<String> files = Stream.of(leftContents, rightContents)
+                .flatMap(contents -> contents.keySet().stream())
+                .sorted()
+                .distinct()
+                .collect(Collectors.toList());
+        for (String file : files) {
+            if (sb.length() > maxTotalDiffSize)
+                sb.append("--- ").append(file).append('\n').append("Diff skipped: Total diff size >").append(maxTotalDiffSize).append("B)\n\n");
+            else
+                diff(Optional.ofNullable(leftContents.get(file)), Optional.ofNullable(rightContents.get(file)), maxDiffSizePerFile)
+                        .ifPresent(diff -> sb.append("--- ").append(file).append('\n').append(diff).append('\n'));
+        }
+
+        return sb.length() == 0 ? "No diff\n" : sb.toString();
+    }
+
+    private static Optional<String> diff(Optional<ZipEntryWithContent> left, Optional<ZipEntryWithContent> right, int maxDiffSizePerFile) {
+        Optional<byte[]> leftContent = left.flatMap(ZipEntryWithContent::content);
+        Optional<byte[]> rightContent = right.flatMap(ZipEntryWithContent::content);
+        if (leftContent.isPresent() && rightContent.isPresent() && Arrays.equals(leftContent.get(), rightContent.get()))
+            return Optional.empty();
+
+        if (left.map(entry -> entry.content().isEmpty()).orElse(false) || right.map(entry -> entry.content().isEmpty()).orElse(false))
+            return Optional.of(String.format("Diff skipped: File too large (%s -> %s)\n",
+                    left.map(e -> e.size() + "B").orElse("new file"), right.map(e -> e.size() + "B").orElse("file deleted")));
+
+        if (leftContent.map(c -> isBinary(c)).or(() -> rightContent.map(c -> isBinary(c))).orElse(false))
```

Comment:
If left is non-binary, and present, but right is binary, this should still trigger, but it won't.

Reply:
Thanks, fixed.